### ISSUE TITLE
Reliable way of exporting complex UIElements to a bitmap

### ISSUE THREAD

**Issue by jroessel** [dotnet/wpf#3682]
One topic that has come up in the previous years a lot from customers of our WPF-based library is that image exports can be unreliable when using XAML-based controls inside. The exhibited problems vary between:

- missing visuals
- misplaced visuals
- bindings not applied
- triggers not run
- values from converters not applied

Last year I've fixed those, or so I thought, by changing how we actually export our images. The previous code path looked basically as follows:

```csharp
// Export visual
var visual = ourControl.ExportEverythingAsVisual();
// Place in container
// container is a custom class, but basically a UIElement that can contain both Visuals and other UIElements
// and passes Measure/Arrange calls to its children, as well as implement our own layout logic, which uses a custom AttachedProperty on our control
container.Add(visual);
// Perform a single layout pass (or so we thought)
container.Measure(new Size(imageWidth, imageHeight));
container.Arrange(new Rect(0, 0, imageWidth, imageHeight));
// write to bitmap
bitmap.Render(container);
// encode and save to file; not relevant here, I guess
```

Upon reading the ContextLayoutManager source code a bit I noticed that there are some things that won't ever work correctly with explicit calls to Measure/Arrange since they defer part of the work into a second layout pass, which only the actual layout manager can resolve and perform. I think things like event handlers on `SizeChanged` can cause this, or `SharedSizeGroups` for grids.

So I then set out to try to fix our export with the knowledge I had at that time. I even finally had a test case by a customer that was simple enough to unit-test to verify that things worked correctly after the fix, whereas they were broken before. The new code basically looked as follows:

```csharp
// Perform a single layout pass - doesn't help much, but shouldn't hurt much either, except performance
container.Measure(new Size(imageWidth, imageHeight));
container.Arrange(new Rect(0, 0, imageWidth, imageHeight));
// Wait for WPF's layout to finish by scheduling at a priority _after_ layout
ourControl.Dispatcher.Invoke(DispatcherPriority.Loaded, new Action(() => {
    // write to bitmap
    bitmap.Render(container);
    // encode and save to file; not relevant here, I guess
}));
```

Turns out, this works most of the time, but apparently not always. One customer performed image exports of different diagrams in a loop and found random elements to be missing from the export. From reading the source code I remembered that the layout manager may [defer remaining elements to a later frame if layout takes too long](https://github.com/dotnet/wpf/blob/d49f8ddb889b5717437d03caa04d7c56819c16aa/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/LayoutManager.cs#L297-L309). What I didn't realize at the time was that those deferred updated are posted at `Background` priority, so waiting for `Loaded` will get all updates from the first 153 elements + 153×2 ms, but would not get the results of the later layouts that are happening at a lower priority.

My current guess would be that things _might_ work when changing `Loaded` above to `ContextIdle`. But it still nags at me that I cannot figure out the »correct« way of exporting UIElements so that all layout has happened at export time. `VisualBrush` somehow manages to wait for all that before painting and so do printing or XPS export, but those have the benefit of being able to call internal APIs.

Is there an official correct way of waiting for layout completion before rendering to a bitmap? Am I on a completely dangerous and wrong path? Almost correct but missing an important detail?

**Comment by ClosetBugSlayer** [comment]
I assume the "bitmap" in the code snippets refers to a RenderTargetBitmap?

**Comment by jroessel** [comment]
> I assume the "bitmap" in the code snippets refers to a RenderTargetBitmap?

Yes. Updated the comments in the code.

**Comment by SamBent** [comment]
First, don't call Measure/Arrange directly;  instead just call `container.InvalidateMeasure()` (after ensuring its `Height` and `Width` are set to the desired values).  This marks your container as "dirty" and causes it to participate in the next layout pass.

Second, there is no guaranteed way to tell when layout is complete.   Each layout pass calls Measure and Arrange on a set of dirty elements, then raises events about the changes that happened.  Any of these calls can introduce new elements or make existing elements dirty, requiring another layout pass.  The "move layout to Background" logic you mentioned is WPF's way of coping with this potentially never-ending process;  it keeps complicated layouts from starving the render thread or the reaction to user input.   

In practice, layouts are rarely that complicated;  waiting for Loaded priority usually works, and ContextIdle should work except in cases that explicitly ask for infinite layout (which, like infinite loops in programs, are impossible to detect in all cases).  Another alternative is to listen for the `LayoutUpdated` event, which is raised near the end of each layout pass.

Third, you can query the `IsMeasureValid` and `IsArrangeValid` properties on your container to see whether its layout is current.  If not, simply call `BeginInvoke` to try again later.

**Comment by jroessel** [comment]
Could it be that `UIElement`s that are not currently shown anywhere do not participate in layout at all? I understand that manually measuring and arranging the `UIElement` isn't necessary, but here it at least ensures that the resulting size is correct. If I just mark it as MeasureDirty the layout doesn't pick it up, not even when calling `UpdateLayout()` (which, as far as I gathered from the source, doesn't really care about the `UIElement` you call it on, so it's rarely the right hammer for any nail, I guess).

My export code is as follows now:
```
      var grid = new Grid {
          Width = configurator.ViewWidth,
          Height = configurator.ViewHeight,
          Children = { canvasContainer }
      };
      grid.InvalidateMeasure();
      grid.UpdateLayout();

      // Wait for layout to complete
      void Action() {
        if (!grid.IsArrangeValid) {
          grid.Dispatcher.BeginInvoke(DispatcherPriority.ContextIdle, (Action) Action);
        } else {
          bitmap.Render(grid);
          bitmap.Freeze();
          BitmapFrame[] frames = CreateBitmapFrames(bitmap);
          encoder.Frames = frames;
          encoder.Save(stream);
        }
      }

      grid.Dispatcher.BeginInvoke(DispatcherPriority.Loaded, (Action) Action);
```

The grid was necessary as the `canvasContainer` isn't really a layout panel and doesn't have a size per se (its ArrangeOverride just returns 1×1).

When I look at the Measure/Arrange queues in the layout manager, they're always empty. Is there a way of forcing the an offscreen visual to get into those queues or is the whole idea basically not a good one?

**Comment by vatsan-madhavan** [comment]
> Is there a way of forcing the an offscreen visual to get into those queues or is the whole idea basically not a good one?

Offscreen in what way? Are you rendering on a headless server, for e.g.?

**Comment by jroessel** [comment]
> Offscreen in what way? Are you rendering on a headless server, for e.g.?

No, but the grid in the code above isn't part of any window that's displayed, and I seemingly couldn't get it to appear in the measure/arrange queue of the layout manager during testing. Perhaps I was just looking at it wrong, though, because some layout does sometimes happen, even with visuals that are not part of anything that's displayed.

**Comment by SamBent** [comment]
Try adding the grid to a convenient place in the visual tree, but setting Visibility="Hidden" to keep it from displaying.  (Convenient means some place where its layout space won't affect anything else.)   

If your container isn't getting added to the measure queue, it won't get all the benefits of the real layout algorithm.  For example, it will see the Measure you call explicitly, but not any subsequent re-measures that may arise due to changes in its subtree (such as a binding providing a new value that changes the size of a TextBlock).
